Sparteo: Add required query params to adapter endpoint

t-sormonte opened the pull request:
### 🔧 Type of changes
- [X] bid adapter update


### ✨ What's the context?
We want to support new required query params when calling our SSP
The Java PR: https://github.com/prebid/prebid-server-java/pull/4225

t-sormonte committed:
Sparteo: add required query params to adapter endpoint

# Conflicts:
#	macros/macros.go

# Conflicts:
#	adapters/sparteo/sparteo.go
#	adapters/sparteo/sparteo_test.go
#	static/bidder-info/sparteo.yaml

ccorbo commented on adapters/sparteo/sparteo.go:
```diff
@@ -162,6 +193,115 @@ func (a *adapter) MakeRequests(req *openrtb2.BidRequest, reqInfo *adapters.Extra
+	return strPtr(b)
+}
+
+func ensurePublisher(p **openrtb2.Publisher) *openrtb2.Publisher {
```
why pass a double pointer? Can pass a single pointer and its clenaer
Example:
```
func ensurePublisher(p *openrtb2.Publisher) *openrtb2.Publisher {
	if p == nil {
		return &openrtb2.Publisher{}
	}
	return p
}
```

usage:
```
req.Site.Publisher = ensurePublisher(req.Site.Publisher)
```

t-sormonte replied:
indeed, updated

t-sormonte committed:
Code Review ccorbo

t-sormonte commented on adapters/sparteo/sparteo.go:
```diff
@@ -162,6 +193,115 @@ func (a *adapter) MakeRequests(req *openrtb2.BidRequest, reqInfo *adapters.Extra
+	return host
+}
+
+func strPtr(s string) *string { return &s }
```
Updated

ccorbo commented on static/bidder-info/sparteo.yaml:
```diff
@@ -13,8 +13,8 @@ capabilities:
       - video
       - native
-endpoint: "https://bid.sparteo.com/s2s-auction"
+endpoint: "https://bid.sparteo.com/s2s-auction?network_id={{.NetworkId}}&{{if .SiteDomain}}&site_domain={{.SiteDomain}}{{end}}{{if .AppDomain}}&app_domain={{.AppDomain}}{{end}}{{if .Bundle}}&bundle={{.Bundle}}{{end}}"
```
won't this result in an extra '&' after network id, also could you just pass the params if they're present or not and they will come thorugh as empty params instead of using the condtionals?

t-sormonte replied:
I wasn't sure if i had to stick with replacement macro building the conditional template dynamically or i could just use query param to mirror my Prebid Java adapter logic. 
I chose to mirror my java logic but if i have misunderstood, please let me know, i'll adjust the code.

ccorbo replied:
yeah you want to use the endpoint template to build your endpoint url with macros

Example here: https://github.com/prebid/prebid-server/blob/master/adapters/adview/adview.go#L163

t-sormonte committed:
Code Review ccorbo 2

t-sormonte commented:
Hi @bsardo , @ccorbo  and @pm-isha-bharti , 
just a quick update: [Java](https://github.com/prebid/prebid-server-java/pull/4225) and [Prebid.js](https://github.com/prebid/Prebid.js/pull/13986) PRs have been merged. 

When you have a moment, could we please continue the review on the Go version so it can stay aligned with the other implementations?

ccorbo commented on adapters/sparteo/sparteo.go:
```diff
@@ -162,6 +190,148 @@ func (a *adapter) MakeRequests(req *openrtb2.BidRequest, reqInfo *adapters.Extra
+	return updated, nil
+}
+
+func (a *adapter) buildEndpointURL(networkId string, siteDomain string, appDomain string, bundle string) (string, error) {
```
Why dont you build the url via the templated endpoint, if any param isn't present use the unknown value. You have a lot of code that could be simplified using this approach

Example here: https://github.com/prebid/prebid-server/blob/master/adapters/adview/adview.go#L163

ccorbo commented on adapters/sparteo/sparteo.go:
```diff
@@ -3,29 +3,41 @@ package sparteo
-		endpoint:   cfg.Endpoint,
-		bidderName: string(bidderName),
-	}, nil
+	templateString := cfg.Endpoint + "?network_id={{.NetworkId}}{{if .SiteDomain}}&site_domain={{.SiteDomain}}{{end}}{{if .AppDomain}}&app_domain={{.AppDomain}}{{end}}{{if .Bundle}}&bundle={{.Bundle}}{{end}}"
```
@t-sormonte 

Sorry if there was confusion here, we don't want conditionals in the template, either use the template without conditionals. Or if you don't want query params that are empty in your URL string, conditonally build the URL without using macros.

t-sormonte commented:
Hi @bsardo @pm-isha-bharti ,
just bumping this thread.

We’d love to get this aligned as soon as possible.

bsardo reviewed (commented):
@t-sormonte please see my last two comments.

t-sormonte committed:
Go fmt after conflict resolution

t-sormonte committed:
Fix import to V4 module

t-sormonte committed:
Fix missing ptrutil import in Sparteo adapter

The ptrutil package was used by domain/bundle resolution helpers
but not imported, causing a build failure.

t-sormonte commented:
Hi @bsardo, i'm keen to get this across the finish line. 
Please let me know if there are any blockers.
Thanks!

t-sormonte committed:
Add GPP query params to Sparteo iframe sync URL

t-sormonte commented:
Hi @ccorbo i have rebased the branch but it cancel you approval. 
Could you please reapprove ? Thanks

t-sormonte commented:
Hi @pm-nikhil-vaidya i'm keen to get this across the finish line.
Can you please review / approve ?
Thanks!

t-sormonte commented:
Hi @ccorbo @przemkaczmarek  
thanks for the review/approvals.
@bsardo could this be merged soon ? Also is there anything else you need from my side ?
Thanks team !
